# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/jamieleeuw/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Lane: Refresh / Content Opportunity Scoring

I'm picking this lane because I've already seen it work end-to-end in the Week 1-2
notebooks: a simple hand-written rule (stale + visible pages) gets real but limited
precision, and a learned model beats it — though the gap shrinks or disappears once
evaluated honestly on a client-holdout split rather than in-sample. That gave me a
concrete feel for the problem before committing to it. The question "which pages
should a reviewer look at first?" — is also the most directly actionable of the four
lanes: it produces a ranked queue a real content team could use, with reason codes
they can inspect, rather than just a descriptive report.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Decision improved: which pages in a content portfolio get reviewed first, given a
content team only has capacity to review a limited number of pages per cycle (e.g.
the top 20-50).

Who acts: a content strategist or SEO reviewer working through a prioritized queue.

The action: review the flagged page and decide to refresh, expand, protect, prune,
or monitor it — the model doesn't take action itself, it just orders the queue.

Cost of a wrong call: two directions.
- False positive (flagged page isn't actually a problem): wastes a reviewer's limited
  time — the page is fine, so the review produces no benefit, though the cost is
  bounded to one reviewer-hour.
- False negative (a genuinely declining, high-value page isn't flagged and gets missed):
  more costly, since real organic traffic keeps eroding unnoticed until someone catches
  it another way.
This asymmetry is why precision@K (are the top K recommendations actually good?) matters
more here than plain accuracy — the review team only ever looks at the top of the list.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [5]:
import pandas as pd
import os

# Walk up until we find the repo root (where data/raw lives)
while not os.path.isdir("data/raw") and os.getcwd() != os.path.dirname(os.getcwd()):
    os.chdir("..")

print("Working dir:", os.getcwd())

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

declining_rate = (df["trend_direction"].str.lower() == "down").mean()
stale = (df["days_since_last_update"] >= 180)
visible = (df["impressions_90d"] >= 500)
stale_and_visible_pct = (stale & visible).mean()

print(f"Rows: {len(df):,}")
print(f"Declining rate (trend_direction == 'down'): {declining_rate:.3f}")
print(f"Share of pages that are both stale (>=180d) and visible (>=500 impressions/90d): {stale_and_visible_pct:.3f}")

Working dir: c:\Users\Japan\flyrank-ml-internship
Rows: 30,000
Declining rate (trend_direction == 'down'): 0.542
Share of pages that are both stale (>=180d) and visible (>=500 impressions/90d): 0.001


Three numbers that make this lane worth pursuing:
1. 30,000 pages in the starter slice alone; the real warehouse has ~520k content
   items and ~79M daily fact rows — plenty of scale for this to be worth 7 weeks.
2. ~54.2% of pages are currently labeled "declining" (trend_direction == down) — a
   large, well-balanced target class, not a rare-event problem.
3. Only ~0.1% of pages are both stale (>=180 days since update) and still visible
   (>=500 impressions/90d) — meaning the naive hand-rule from Week 1-2 flags a very
   narrow slice of the inventory. That's exactly why a learned model matters here:
   the simple rule is too conservative to surface most of the real opportunity, and
   the verified pipeline results back this up — Precision@50 = 0.240 for the baseline
   rule vs. 0.740 for a random forest on the same client-holdout split.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.



What this work can say: which pages, based on observed 90-day search and content
signals, look most similar to pages that later showed declining trends in this
dataset — an observed, directional, decision-support ranking meant to help a human
reviewer prioritize limited time.

What this work cannot say: that any specific page will decline, that refreshing a
page will cause it to recover (that needs a real experiment, not this data), or
anything about how Google's ranking algorithm actually works. "Declining" here is a
proxy label built from a trend bucket in the current window — not a validated
future-looking outcome yet. A stronger version of this lane, which I may build toward,
would define the label as a future 30-day outcome measured after a 90-day feature
window, with a proper leakage audit — closer to what real decision-support needs.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.